src/Bootstrappers/PrefixCacheTenancyBootstrapper.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Bootstrappers;
+
+use Illuminate\Cache\Repository;
+use Illuminate\Support\Facades\Cache;
+use Stancl\Tenancy\Contracts\TenancyBootstrapper;
+use Stancl\Tenancy\Contracts\Tenant;
+
+class PrefixCacheTenancyBootstrapper implements TenancyBootstrapper
+{
+    protected null|string $originalPrefix = null;
+    protected string $storeName;
+
+    public function bootstrap(Tenant $tenant): void
+    {
+        $this->originalPrefix = config('cache.prefix');
+        $this->storeName = config('cache.default');
+
+        $this->setCachePrefix('tenant_id_' . $tenant->id);
+    }
+
+    public function revert(): void
+    {
+        if ($this->originalPrefix) {
+            $this->setCachePrefix($this->originalPrefix);
+            $this->originalPrefix = null;
+        }
+    }
+
+    protected function setCachePrefix(string $prefix): void
+    {
+        config()->set('cache.prefix', $prefix);
+
+        app('cache')->forgetDriver($this->storeName);
+
+        // This is important because the `CacheManager` will have the `$app['config']` array cached
+        // with old prefixes on the `cache` instance. Simply calling `forgetDriver` only removes
+        // the `$store` but doesn't update the `$app['config']`.
+        app()->forgetInstance('cache');
+
+        //This is important because the Cache Repository is using an old version of the CacheManager
+        app()->forgetInstance('cache.store');
+
+        // Forget the cache repository in the container
+        app()->forgetInstance(Repository::class);
+
+        Cache::clearResolvedInstances();
```

Comment:
Do we need **all** of these 4 calls?

Reply:
The current test is passing without these four calls. 
But I think we should keep the `Cache::clearResolvedInstances();` one and remove the other three.

Reply:
So after adding the new test suite, a test failed if we remove `$this->app->forgetInstance('cache.store');`. For now, I've removed the following. 
```php
$this->app->forgetInstance('cache');

$this->app->forgetInstance(Repository::class);
```

And current tests are passing.

Reply:
We should know very clearly what each binding is used for and if it needs to be cleared, rather than trying to randomly clear bindings. If you miss some edge case in tests, there will be bugs, so it'd be better to be more intentional about what the code is doing here.

Reply:
> We should know very clearly what each binding is used for and if it needs to be cleared, rather than trying to randomly clear bindings. 

I agree. So I researched these bindings and why the Spatie package adds these. 

### Reset Facade cache `Cache::clearResolvedInstances()`
It is needed when a call to the facade has been made before bootstrapping tenancy. The facade has its own cache, separate from the container. That's why we need to clear facade cache. 

### Forget cache store manager `$this->app->forgetInstance('cache.store')`
It is needed because the Cache Repository uses an old version of the CacheManager so we need to forget it. Here is the PR someone sent to Spatie package and described the issue and reason for adding this https://github.com/spatie/laravel-multitenancy/pull/374

### Forget cache instance `$this->app->forgetInstance('cache')`
Simply calling `forgetDriver()` isn't enough to update cache prefixes everywhere. The `CacheManager` also keeps a copy of the old config in `cache`.  These should be cleared and forgotten after switching cache configs. Here is the related PR https://github.com/spatie/laravel-multitenancy/pull/355

### Forget cache repository in the container `$this->app->forgetInstance(Repository::class)`
I'm not really sure if we needed this. Spatie package added this recently. 
<img width="950" alt="image" src="https://user-images.githubusercontent.com/54532330/203483131-890b5653-aac4-4dd0-a3f1-4e2a04e351e6.png">
Don't you think forgetting the `cache.store` should also forget the `Repository`? 


So currently, I'm not sure if we should keep all these calls. As I added a separate test suite, I'm thinking of edge cases for tests. And maybe there, I can actually feel the need for these calls. 
But from the Spatie package PRs, it looks like people faced issues, so they contributed. We might also want to keep all these 4 calls.

Reply:
Thanks a lot for the detailed write-up.

I guess we can keep all of them then, just add a concise comment above each of those lines explaining in short what it's used for/why it's needed.

And regarding `Repository`, I guess we can just keep it there and mention in the comment that it's added just to cover edge-cases.

I guess you don't need to cover **low-level things** in tests if they work well in the Spatie package.

But we likely still want some specific tests for this bootstrapper that make sure it doesn't affect other caches, that central cache still works well, that everything works well after tenancy is ended (and also ended & reinitialized) etc. Just try to cover *usage edge-cases*, and stuff that's specific to the implementation that you think should be tested. Look at the other bootstrapper tests for inspiration.

Reply:
Ok, I added all four calls and added the comments on each call. 

As for the tests, I added a few tests (which I can think of). Most of them are inspired by the existing bootstrapper test suite.